File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {

    let set1 = new Set();
    let maxCount =0;

    let left =0
    let right =0
    while((left <=right && right <= s.length-1)){
        let count =0
        while(!set1.has(s[right])){
            //console.log(s[right])
            set1.add(s[right]);
            if(right <s.length-1)
                right++;
            else
                break;
            //console.log("inside",right)
        }
        
        maxCount = Math.max(maxCount,set1.size);

        set1.delete(s[left]);
        //console.log(set1,right,left)

        if(right >= s.length-1)
            break;
        else
            left++;
        
    }
    
    return maxCount;
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        map1 = {}
        left =0
        c = 0

        for right in range(len(s)):

            if s[right] in map1 and map1[s[right]] >= left:
                left = map1[s[right]]+1

            map1[s[right]] = right
            c = max(c, right - left +1)
        
        return c

        